### PR TITLE
[Mega PR] Implement Scratch Auth and few fixes

### LINKED ISSUE #19: make sure typeof string

### LINKED ISSUE #26: use req.get("Host") instead of localhost:8081

### PR DESCRIPTION
**Pull request by Weredime**
Resolves #26 
Resolves #19

I tried testing, but my dev mongo database isn't functioning correctly; somebody with a proper working database please test

# Changes

- [x] Implements Scratch Auth
- [x] Use constants instead of let/var.
- [x] Stops ocular from 500ing when ScratchDB is down

cc @Looky1173 just want to see if the implementation is correct

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,10 +1,12 @@
 require('dotenv').config()
-const fetch = require('node-fetch')
-const crypto = require('crypto')
+const crypto = require('crypto');
 
+const cors = require('cors');
+const fetch = require('node-fetch');
 const express = require('express');
-const db = require('monk')(process.env.MONGO_URL || 'localhost/my-ocular-other')
```
Comment (Looky1173) [inline index.js:6]: I think this change is not necessary.

File: index.js [hunk 2]
```diff
@@ -1,10 +1,12 @@
 require('dotenv').config()
-const fetch = require('node-fetch')
```
Comment (Looky1173) [inline index.js:2]: Don't remove the `node-fetch` package! [Only Node.js v17.5 and above support `fetch` natively.](https://nodejs.org/tr/blog/release/v17.5.0/)

File: index.js [hunk 3]
```diff
@@ -326,52 +331,52 @@ app.post('/api/reactions/:id', cors(corsOptions), async (req, res) => { // react
...
-        res.redirect(`https://fluffyscratch.hampton.pw/auth/getKeys/v2?redirect=bG9jYWxob3N0OjgwODEvYXV0aC9oYW5kbGU=`)
-    } else {
-        res.redirect(`https://fluffyscratch.hampton.pw/auth/getKeys/v2?redirect=bXktb2N1bGFyLmplZmZhbG8ubmV0L2F1dGgvaGFuZGxl`)
-    }
+    const redirectURL = `${req.protocol}://${req.get('host')}/auth/handle`;
+    const encodedRedirectURL = Buffer.from(redirectURL).toString('base64');
+
+    res.redirect(307,
+        `https://auth.itinerary.eu.org/auth/?name=ocular&redirect=${encodedRedirectURL}`
+    )
```
Comment (Looky1173) [inline index.js:345]: This seems fine, but please _do_ test your changes.

File: index.js [hunk 4]
```diff
@@ -326,52 +331,52 @@ app.post('/api/reactions/:id', cors(corsOptions), async (req, res) => { // react
...
-
-        if (redirect != 'localhost:8081/auth/handle' && redirect != 'my-ocular.jeffalo.net/auth/handle') {
-            return res.send('invalid redirect')
+        if (redirect !== redirectURL) {
+            return res.send('invalid redirect') // todo: frontend
         }
 
-        let scratchResponse = await fetch(`https://api.scratch.mit.edu/users/${authData.username}/`)
-        let scratchData = await scratchResponse.json()
+        const scratchResponse = await fetch(`https://api.scratch.mit.edu/users/${authData.username}/`, {
```
Comment (Looky1173) [inline index.js:363]: `authData` is not defined in this scope!

**Review by Looky1173 — CHANGES_REQUESTED** [review]
@Weredime I cannot approve of these changes until they are tested. If your local MongoDB installation is not working properly, try [MongoDB Atlas](https://www.mongodb.com/cloud/atlas/register). Also, I don't understand the need for the `typeof` checks when `!expression` is perfectly suited for most empty value checks; in fact, I heard that `typeof` checks are slower than alternatives.

I think that instead of making a "mega PR", you should submit your changes in smaller, individual PRs to facilitate testing and code reviews.

File: index.js (hunk 2 above)
Comment (Weredime) [reply index.js:2]: Didn't see that change 😅

File: index.js (hunk 1 above)
Comment (Weredime) [reply index.js:6]: This is just to make the imports a bit cleaner.

**Comment by Weredime** [comment]
Superseded by #34.

**Pull request closed without merging**